Initial PLD and Auto Attack Events Implementation

namm100 opened the pull request:
Hello!

Here is a pass at PLD (with 7.2 potency changes). 
In order to model oath gauge generation, I've added a slew of functions/methods/fields pertaining to this.

The main mechanic is just a simple recurring event called `"aa tick"` that is continuously added, removed, modified depending on the actions taken by the user. Most of the action is done in `Game/GameState.ts`

For skills/abilities/limit break??, I've modified `Game/Skills.ts` to include a `startsAuto` parameter that one can modify for any individual skill/ability.
By default:
Weaponskills -> `true`
Spells -> `false`
Abilities -> `false`
Limit Break -> `false`

I've defined two resources called `"AUTOS_ENGAGED"` and `"STORED_AUTO"`. Booleans that track if the auto attack timer is ticking and if the auto attack timer is full but the gauge isn't ticking respectively.

Back to `Game/GameState.ts`, I've defined two methods. First is `onAutoAttack()` that jobs can override to define what happens on an auto attack. Second is `onAuto()` which is supposed to add a damage event for the auto and then call `onAutoAttack`. The damage event is not implemented yet.

The `"aa tick"` event is modified when a weaponskill/spell or an ability is pressed. Depending on a combination of states such as if the ability has a cast time, user is in combat, autos are ticking, etc, the recurring event is removed and restarted to fit that particular state.

Finally, I've added a skill to PLD alone (for now) that toggles on and off auto attacks to model if the player right-clicked to toggle autos in-game. Shanzhe wanted this to be in a similar location to positionals, but for now its a skill with no animation lock.

Everything as of now is a work in progress but at least (i think) PLD itself is functional with auto attacks to generate oath gauge.

The console is popping out "Unexpected value NaN parsing x2/y2 attribute" for whatever reason as well. Maybe this has to do with the UI as I've added a lot of timers for PLD specifically.

Thanks,

namm100/Sterling

miyehn commented:
@zqsz-xiv (I'm assuming you two are in contact but just in case 👀)

zqsz-xiv commented:
planning to take a look after savage crunch (hopefully i'll clear fast)

zqsz-xiv commented on src/Game/GameState.ts:
```diff
@@ -640,6 +651,114 @@ export class GameState {
+			this.removeAutoAttackTimer();
+		} else if (this.resources.get("STORED_AUTO").available(1)) {
+			// do an auto attack on a delay according to castTime
+			const event = new Event("stored auto", castTime ? castTime : 0, () => {
```
what happens if an auto event would have occured in the middle of a cast bar?

zqsz-xiv replied:
I believe this code would fail if an auto-attack is performed at time _t_, and the next auto at _t + 3_ (assuming base sks) occurs in the middle of a castbar. I attached an example to the top-level review comment.

zqsz-xiv commented on src/Game/GameState.ts:
```diff
@@ -665,12 +784,88 @@ export class GameState {
+		const currentDelay = this.findAutoAttackTimerInQueue();
+		const startsAutos = skill.startsAuto; // <<---  for spells starting autos (eg. RDM)
+
+		/* console.log(
+			"Using: " + skill.name + " cast: " + capturedCastTime + " starts auto: " + startsAutos,
+		);
+		*/
```
```suggestion
```

zqsz-xiv commented on src/Game/GameState.ts:
```diff
@@ -665,12 +784,88 @@ export class GameState {
+			state.startAutoAttackTimer(aaDelay);
+			*/
+
+			if (this.isInCombat()) {
```
can you put this block of logic into a separate function so it's easier to read?

zqsz-xiv reviewed (commented):
Thanks for the work. I haven't checked specific skill potencies/job levels yet, but I'll do so in a later review.

The console error seems to be because you didn't declare a cooldown length for `cd_AUTO_ATTACK`, so it doesn't know what to do when trying to draw the cooldown spinner overlay for the skill icon. Either declare a fake cooldown as something really small (like 0.01s), or use the positional-like buff toggle instead.

namm100 commented on src/Components/Asset/Buffs/PLD/Guardian's Will.png:
I always forget which is which, but pressing Guardian activates both the 40% damage reduction AND the shield.

namm100 commented on src/Components/Asset/Buffs/PLD/Goring Blade DoT.png:
Comeonnnnnnnnnn :)

namm100 commented on src/Components/Asset/Buffs/PLD/Divine Veil huh.png:
Great question! Maybe PvE but I'm not sure.

namm100 replied:
Actually, I think it's old Divine Veil that proc-ed only on a heal.

namm100 commented on src/Components/Asset/Buffs/PLD/Blade of Valor DoT.png:
Welllllllll hahaha yea not used anymore sad :(

namm100 commented on src/Game/Jobs/PLD.ts:
It is the intended behavior. In game I'm pretty sure pressing any of paladin's ranged gcds enables auto attacks (like right-clicking). However, its true that you'd be pressing shield lob if you are far away so autos wouldn't start ticking. Right now, autos work assuming the PLD is always in melee range. 

There is of course the option to just disable starting auto attacks for shield lob and other ranged gcds and make the user have to re-enable it. Or we could just have shield lob never enable auto attacks.

This is mostly a design choice. I personally think that it should be this way, but make the user have to disable autos to simulate moving away from the boss.

zqsz-xiv replied:
When I wrote the initial comment I misread the logic and didn't realize that spells like HS and Blade's Combo also initiate auto-attacks as soon as they're pressed.

With this in mind, I think it makes more sense to force users to manually re-enable auto-attacks after using disengage tools (whether it's Shield Lob, HS, or a Blade's ability). Otherwise, modeling extended disengage periods that last for multiple GCDs would force the users to do something like this:
- melee weaponskil
- toggle autos [off, disengage starts here]
- HS
- toggle autos [off]
- imperator
- toggle autos [off]
- blade of whatever
- toggle autos [off]
- intervene [autos automatically resume, disengage ends]
- blade of whatever

which could get very annoying. If we only re-enable autos with melee-ranged skills, the process instead would be
- melee weaponskill
- toggle autos [off, disengage starts here]
- HS
- imperator
- blade of whatever
- intervene [autos automatically resume, disengage ends]
- blade of whatever

The line gets kind of blurry for stuff like Reprisal that can technically be pressed outside melee range, but always start autos in-game. I do think at least for the extremely common cases (ranged filler and spells like HS and Blade's), it is worth enough of a benefit to user convenience to add an exception that contradicts the actual in-game rules.

namm100 commented on src/Game/GameState.ts:
```diff
@@ -170,6 +171,8 @@ export class GameState {
 		this.skillsList = new SkillsList(this);
 		this.displayedSkills = new DisplayedSkills(this.job, config.level);
+
+		this.autoAttackDelay = 2.5; // defaults to 2.5 can be changed by referencing this field
```
Yes but riddle of wind....... hmmm

zqsz-xiv replied:
If we want to account for AA haste buffs in the future, then the calculation function should just check whether the buff is active when starting the next auto timer rather than needing to update and track a second-order state variable.

namm100 commented on src/Components/Asset/Buffs/PLD/Requiescat Zero.png:
Needed for buff icon in timeline.

namm100 commented on src/Components/Asset/Buffs/PLD/Blade of 2.png:
I realized this is PvP only so I'll remove it.

zqsz-xiv replied:
oh, are these tracked as an actual combo instead of by buff in pve?

namm100 commented:
Just pushed an initial response to the comments.
The only thing not addressed is the movement of the auto attack button to a place similar to positionals. 
I'll try to see if I can do this.

namm100 commented on src/Game/GameState.ts:
```diff
@@ -640,6 +651,114 @@ export class GameState {
+	findAutoAttackTimerInQueue() {
+		// hi
+		let timer = -1;
+		this.eventsQueue.forEach((event) => {
```
I kept the `forEach` loop here because I don't mark the stored auto event (event that when times out you get a stored auto buff) with a tag. If that's okay.

namm100 commented:
UI change. Removed the auto attack 0/1 thingy. Added an auto attack toggler in the spot where positionals are for melees. For now I just resized the auto attack icon. A different icon can be added, if needed. 
<img width="951" alt="Screenshot 2025-04-22 at 7 41 21 PM" src="https://github.com/user-attachments/assets/b8879b95-3ea1-452c-a72c-f9b713ca17a2" />

zqsz-xiv commented on src/Game/GameState.ts:
```diff
@@ -640,6 +658,226 @@ export class GameState {
+	 * Function to start auto attacks
+	 * removes old auto attack timer
+	 * starts a new recurring auto attack timer
+	 * timer for auto initial + recurringDelay(defaults 3)
```
can you add to the comment explaining why we need separate values for `initialDelay` and `recurringDelay`?

namm100 replied:
I'll write something similar here too.
I call this function many times to overwrite the auto timer. Initial delay basically for tracking the old event and also including a cast time or some sorta break in the flow.

zqsz-xiv commented on src/Game/GameState.ts:
```diff
@@ -669,8 +907,11 @@ export class GameState {
-		// create a potency object so an aggro-only potency mark can be drawn.
-		if (!potency && (potencyNumber > 0 || skill.drawsAggro)) {
+		// create a potency object so a damage mark can be drawn if we're not already in combat.
+		if (!potency && (potencyNumber > 0 || (skill.drawsAggro && !this.isInCombat()))) {
```
Can you move the auto start logic to a separate if/else statement from the potency mark drawing? I think we want to keep the behavior of the orange aggro mark being drawn even if you're already in combat (so the application delay of aggro-only skills is still shown).

zqsz-xiv commented on src/Game/Jobs/PLD.ts:
```diff
@@ -1,74 +1,1162 @@
+makePLDResource("DIVINE_VEIL", 1, { timeout: 30 });
+makePLDResource("ATONEMENT_READY", 1, { timeout: 30 });
+makePLDResource("DIVINE_MIGHT", 1, { timeout: 30 });
+makePLDResource("INTERVENTION", 1, { timeout: 8 }); // TODO??
```
?

namm100 replied:
Just like passage of arms below I wasn't sure whether to include it or not as party members receive the buff and not the PLD itself.

zqsz-xiv commented on src/Game/Jobs/PLD.ts:
```diff
@@ -1,74 +1,1162 @@
+makePLDResource("REQUIESCAT", 4, { timeout: 30 });
+makePLDResource("CONFITEOR_READY", 1, { timeout: 30 });
+makePLDResource("PASSAGE_OF_ARMS", 1, { timeout: 18 });
+makePLDResource("ARMS_UP", 1); // ???
```
Is this the buff granted to party members by passage?

namm100 replied:
see above comment.

zqsz-xiv commented on src/Game/Jobs/index.ts:
```diff
@@ -33,7 +33,6 @@ export function getGameState(config: GameConfig): GameState {
 			return new DRKState(config);
 		case "GNB":
 			return new GNBState(config);
-
```
add back the line for consistency

zqsz-xiv reviewed (commented):
A couple more auto-related things (see also my sub-comment about spells resuming autos)
- In the current implementation, it looks like no auto-attack happens at t=0 (or whatever the application delay of an auto is). I think there should be some consideration to allow users the option to do a macro pull so autos don't start until first ability damage application, but we can deal with that later and just do a naive "start ASAP" for now.
- An auto that is scheduled to occur in the middle of a castbar (but started before the castbar was initiated) seems to not go off (see attached file)
[bad_auto_timing.txt](https://github.com/user-attachments/files/19920939/bad_auto_timing.txt)

It would also be helpful if you could transcribe a partial log for testing, something like M7S's first couple of arenas, or any fight where you know there's really tight oath gauge management would be helpful.

zqsz-xiv commented:
@namm100 don't forget to update your branch with `main` by the way, I don't think there were any major changes since you last pulled but you should merge just to be safe.

namm100 commented on src/Components/Jobs/PLD.tsx:
```diff
@@ -1,22 +1,99 @@
+
+		const basicCombo = resources.get("PLD_COMBO_TRACKER");
+		const aoeCombo = resources.get("PLD_AOE_COMBO_TRACKER");
+		// const confiteorCombo = resources.get("PLD_CONFITEOR_COMBO_TRACKER");
```
Probably not? I'll delete it. I guess we'll only track the 1-2-3 and AOE 1-2 combos, otherwise we'd have to make another timer for confiteor :(

namm100 commented:
Pushed the latest updates.
Notable changes:
made `autoAttackDelay` private
made Shield Lob, Holy Spirit, entire Confiteor combo and the followup off global specifically not start autos.

zqsz-xiv commented:
Did you get a chance to make a test fight timeline to make sure everything tracks? Otherwise looks good to me, and I'll merge if you're OK with it.

namm100 commented:
Sure! Ready to merge. Here is a full fight timeline for m7s. A bit random but an example.
[PLD_m7s_example.txt](https://github.com/user-attachments/files/20137658/PLD_m7s_example.txt)

zqsz-xiv commented:
Awesome, thanks for the work!
